### ISSUE TITLE
Newer bob versions are slow with many root recipes

### ISSUE THREAD

**Issue by kolewu** [BobBuildTool/bob#177]
We still used bob 0.8.0 because the newer versions are very slow. We have 134 root recipes and `bob ls` lets us wait 7-8 seconds everytime something has changed. Since the helpers like bash completion don't use a configuration, there often are changes that lead to rehashing and that takes time.

When disabling most of the root recipes, the hashing time drops down to under a second. Is there any other way to cut the hashing times down than reducing the root recipes? In my desperation I plan a script that deactivates all root recipes except for one or two ...

**Comment by jkloetzke** [comment]
How much faster is 0.8 in your setup? Can you share the "time bob ls" for both versions? (You have to delete `.bob-packages*`starting with 0.9 to force a re-calculation of the packages. Deleting `.bob-cache*` will add the yaml parsing overhead but that should not have changed.)

I guess the first problem might be that the bash completion does not pass configuration files to "bob ls". Fixing this might prevent delays between bash completion and Bob invocations in your case.

The current behavior is more optimized towards projects with few root recipes but deep dependency hierarchies where the same packages are used often. I guess that is not the case for you. There is another cache for the data that "bob ls" is using. This only pays off though if Bob doesn't have to re-parse the recipes every time.

May I'll add a `--debug=prof` switch to run the command trough "cProfile" to get some performance numbers. But I do not expect to see something easily fixable.

**Comment by jkloetzke** [comment]
I've added a profile switch. Maybe running the following reveals something:

```sh
bob ls
rm -f .bob-packages* .bob-tree*
bob --debug=prof ls
```

**Comment by jkloetzke** [comment]
Please have a look at #178 . Maybe it improves your situation already a bit. The bash completion was really broken with respect to config files or manually passed variables (`-D FOO=bar`).

**Comment by kolewu** [comment]
Profiling: Why do you only remove `.bob-packages*` and `.bob-tree*`? What about `.bob-cache.shelve`?

**Comment by jkloetzke** [comment]
The `.bob-cache.shelve` holds the parsed yaml files (yaml -> python data structures). This overhead did not change between versions and shouldn't make a difference here. This would only pollute the profiling data.

**Comment by kolewu** [comment]
I have worked on profiling the problem. We have another repo with 250 root recipes that needs whopping 86 seconds to build a simple recipe, that only echoes the build states.

To elaborate further on this, I have made a script that removes the *rootness* from all recipes despite the simple one and then adds the original recipes back one at a time. On every step the simple recipe is build after deleting the cache and build dir.

![profile](https://user-images.githubusercontent.com/95135/33676148-2c52dc76-dab5-11e7-9eac-98c3f058c8cf.png)

x is the number of active root recipes and y the build time for the simple recipe.

The big jumps are recipes with many root recipes (all different version modeled as multipackages) **and** many dependencies because these are the main applications.

The script can be found in [this gist](https://gist.github.com/kolewu/7b649acdbc4cd93320f9d8ef4b073e95)

**Comment by jkloetzke** [comment]
That looks interesting. It does not sound completely unexpected, though. In theory old versions of Bob had a run-time roughly equivalent to the number of possible paths in the dependency graph. For our recipes the old versions would easily exhaust the memory because of this behavior.

Newer versions have a run-time of n*m where n is the number of packages and m is the mean number of variants per package. Bob tries to detect duplicates while traversing the dependency graph but this detection comes at a price and only pays off if many equivalent variants of the packages are found.

So in any case there should be an increase in run-time if the number of possible packages and their variants increases. Could you share the pstats file of the worst case? There must be an unexpected hot spot somewhere...

BTW, how does Bob 0.8 perform on the above data set?

**Comment by kolewu** [comment]
With all 250 root recipes activated Bob 0.8 builds the simple recipe in 11,38 seconds.

**Comment by kolewu** [comment]
Since I'm not sure that no personal data is included in the stats file, I can only provide the generated png files for these cases:
* the one before the worst case (116)
* the worst case (139)
* the maximum (250) for the current version and bob 0.8

![bob-recipes-116](https://user-images.githubusercontent.com/95135/33711499-84ca226c-db44-11e7-9fc8-eab468e67090.png)

![bob-recipes-139](https://user-images.githubusercontent.com/95135/33711500-84f98660-db44-11e7-953c-5c23bd09eb6a.png)

![bob-recipes-250](https://user-images.githubusercontent.com/95135/33711501-851418f4-db44-11e7-9d37-05f21e1eb299.png)

![bob-recipes-250_0 8](https://user-images.githubusercontent.com/95135/33711503-852e85d6-db44-11e7-8a68-9fadb3aec093.png)

**Comment by kolewu** [comment]
One of the hotspots is this line: [bob/input.py#L1694](https://github.com/BobBuildTool/bob/blob/master/pym/bob/input.py#L1694). What is the reason to call the `PackageMatcher::matches()` function with copies of `inputEnv` and `inputTools`?

By removing these calls, the execution time drops to a half!

And by fully removing the test for a matching package, the execution time goes to a quarter.

What? | execution time
---------|-------------------
current code | 65,99 sec
without `detach()` | 27,68 sec
without reused packages | 15,12 sec

What is the use-case where this search gives better results?

**Comment by jkloetzke** [comment]
This is exactly the same thing that I was just looking at. This code is there to find already computed identical packages and re-use them instead of calculating them again. We have many recipes with many dependencies between them but with only a small number of variants. So in our case the code at this spot will almost always quickly find a matching package. Without this code Bob would exhaust the available memory in our case.

The `detach()` is necessary because to reliably compute what can be reused Bob has to track the used environment variables in the recipe and its dependencies. The `detach()` gets the variables without marking them as "all used". But the code is really sub-optimal. I'll create a merge request with the first round of fixes.

**Comment by kolewu** [comment]
#184 nearly halves the execution time to 37,81 seconds.

**Comment by jkloetzke** [comment]
:thinking: I had hoped this is already near ~30s. I've amended #184 because the env/tool dicts were still needlessly copied. What did you change exactly for the

> without `detach()`	27,68 sec

case from your measurement?

**Comment by kolewu** [comment]
I simply removed the `detach()` calls from the `matches()` arguments:

```diff
modified   pym/bob/input.py
@@ -1691,7 +1691,7 @@ class Recipe(object):
                 inputTools=Env(), stack=[]):
         # already calculated?
         for m in self.__corePackages:
-            if m.matches(inputEnv.detach(), inputTools.detach(), inputStates, inputSandbox):
+            if m.matches(inputEnv, inputTools, inputStates, inputSandbox):
                 if set(stack) & m.subTreePackages:
                     raise ParseError("Recipes are cyclic")
                 reusedPackage = p = Package()
```

And regarding the *mark as used* functionality: Will it not be triggered by copying, too? Or is the usage counter only in the getter functions that will not be used by copy?

**Comment by kolewu** [comment]
The `MappingProxyType` doesn't change anything in the execution time, it's still around 38 seconds.

So I would prefer to have a configuration option to switch off this reuse feature, since it doesn't have any benefit for our use cases.

**Comment by jkloetzke** [comment]
The `bob.stringparser.Env` class tracks all read accesses. The `detach` method returns a copy that is not tracked in this regard. The new `inspect` method will just return a read-only view without actually copying the data which should further lower the overhead.

I'm a bit reluctant to add an option that exposes this internals. I would rather first try to optimize the feature. Currently the search is O(n^2) where n is the number of variants of a package. I think it should be possible to bring this down to O(log n) in the best case.

**Comment by jkloetzke** [comment]
I tried to create a repo case but didn't succeed so far. I wrote a [script](https://gist.github.com/jkloetzke/e9d625194627c9c52f2718ed8cf3fcba) that generates a hierarchy of recipes and many root recipes. But the run time does not show such a dependency on the re-use feature.

Could you elaborate a bit what is the difference between the multiPackages where the big jump happens? Do they define many variables or only a few? How much are different? Do they define different tools? Right now I have no idea how to reliably trigger the problem.

Nonetheless I updated the pull request with a hackish implementation of a better search algorithm. Could you give it a try?

**Comment by kolewu** [comment]
The real problem is our path-formatter! When I disable it, a simple `bob ls` goes down from 23,54 to 1,58. Sad ☹️.

The pathformatter isn't very sophisticated, but uses its own properties: [path-fmt.py](https://gist.github.com/kolewu/04e96bd470815afb00c955a1067bd3a0)

**Comment by jkloetzke** [comment]
Just to double check: the 23,54s (and 1,58s) are with the latest version of #184, correct?

I could reproduce that enabling the plugin and adding a `checkoutDir` to every recipe already doubles the parsing time. But I don't have an idea yet how this stark difference can be caused by the plugin. The plugin does nothing wrong in this regard. Could you make two more profilings (one with and one without the plugin)? Maybe the diff between the two will give a hint...

**Comment by kolewu** [comment]
Yes, the times are with the latest from your branch `speedup`.

Here are the profile graphs from this branch:

All 250 recipes with *path-fmt* plugin activated in `config.yaml`
![bob-recipes-250__with_path-fmt](https://user-images.githubusercontent.com/95135/34164973-4c78148c-e4db-11e7-9b06-a51aecd30840.png)

All 250 recipes without *path-fmt* plugin activated in `config.yaml`
![bob-recipes-250__without_path-fmt](https://user-images.githubusercontent.com/95135/34164979-4fcd2b18-e4db-11e7-8941-a358263646f1.png)

**Comment by jkloetzke** [comment]
These graphs look like `.bob-packages*` and `.bob-tree*` were not removed before the profiling? Because there is no parsing visible in the graph. :thinking: 

Does the long waiting time happen every time with the plugin enabled? I'd expect it happens only once  and the next invocation with the same parameters is fast, right? But that's not really visible in the above graphs...

**Comment by kolewu** [comment]
No, the cache is removed before every profile run: see [the script](https://gist.github.com/kolewu/7b649acdbc4cd93320f9d8ef4b073e95#file-profile-roots-sh-L22).

And the long waiting time is only once for every cold cache, the second run with hot cache has a time in the range of 1 second.

**Comment by jkloetzke** [comment]
AFAICT both profiles above were generated with 369f384 whereas the current head of the speedup branch is 9b13a39. At least the line numbers in the graphs match the real locations in the source files of the old commit.

I'm also quite sure that the graphs are from invocations where the cache was used and no real parsing was done. The interesting call graph looks like:

build:1173:touch -> pathspec:832:getRootPackage -> input:2491:__generatePackages -> pickle.Unpickler:load

This call path is exactly the expected one for a cache hit. As you can see in the old graphs above "input:2491:__generatePackages" should call "input:1690:prepare" where the bulk of the parsing and calculation is done. The script looks good but something must have gone wrong.

**Comment by kolewu** [comment]
Bob was definitely on the speedup branch at d6c9949d55305b1b806795db341ef41312941f7b only missing the last commit. But there was an error in the script: to also get the timing with hot cache, I added a second run, but forget to change the filename of the stats-file. So the profilings are with hot cache as you suspected.

Here are the runtimes:

path-fmt | cold | hot
-----------|-------|-----
yes  | 81,13 | 6,68
not | 2,33 | 0,98

And here the profiles (only with cold cache):

All 250 recipes with path-fmt plugin activated in `config.yaml`:
![bob-recipes-250__with_path-fmt](https://user-images.githubusercontent.com/95135/34302373-d373ada0-e730-11e7-83a5-49d294eb19ce.png)

All 250 recipes without path-fmt plugin activated in `config.yaml`:
![bob-recipes-250__without_path-fmt](https://user-images.githubusercontent.com/95135/34302401-eb2e0666-e730-11e7-9bc9-5de2340fd2fb.png)

**Comment by jkloetzke** [comment]
I literally took me days to get an understanding what really goes wrong here but I think I finally found the culprit. The root cause is that I internally compare the `PluginState` objects but they are copied for every package. The problem is that no `__eq__` method is implemented an thus Python falls back to object identity. But of course this will never be `True` because we create a unique copy for every possible package.

The solution is quite simple (see 7bbafd00605112a27dc88b5fdf5b6703b3bb41a6). I've updated the speedup branch with some more fixes an it's now at 41146ec299e287b161467ecda6ae0adf666667d2. Could you give that a try?

**Comment by kolewu** [comment]
🥇 This is a great change! With our 250 recipes repo and using the speedup branch on 41146ec I get these times:

action | cache | time
--------|---------|----
build | cold | 2,86 s
build | hot | 1,19 s
ls | cold | 2,27 s
ls | hot | 0,25 s

The cold cache is created by `rm -rf .bob-{packages,tree}* dev` and the *path-fmt* plugin is activated.

**Comment by jkloetzke** [comment]
:sweat_smile: That's even better than I thought. Good that we finally nailed it down. I will polish the PR a bit more and merge it in the next day or two...
